add hack/sync-bug-triage-github-project-beta.sh

helayoty opened the pull request:
#### What type of PR is this:
/kind feature

Fixes: https://github.com/kubernetes/community/issues/6742 
Fixes:  https://github.com/kubernetes/org/issues/3357

#### What this PR does / why we need it:
Duplicate #1968 for Bug Triage Project Beta.

The PR adds a script called `add hack/sync-bug-triage-github-project-beta.sh`.

This script will be run as part of a periodic Prow Job that will sync all open KEP issues from the k/k repository under a current release milestone (say v1.25) to a Bug Triage GitHub Project Beta board (similar to https://github.com/orgs/kubernetes/projects/80).

cc: @Priyankasaggu11929 @kubernetes/release-team-leads @reylejano @cici37 @mrbobbytables

marosset commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,225 @@
+echo 'Startig sync...'
+echo -e "[INFO] Fetching the list of open issues and PRs from k/k under the current release milestone, ${MILESTONE}"
+
+## Fetch issues
```
This query won't be paginated.
You'll want to specify the `--paginate` flag and have the query take in an `$endCursor` arg.

ex:
``` json
$ gh api graphql --paginate -f query='
  query($endCursor: String) {
    viewer {
      repositories(first: 100, after: $endCursor) {
        nodes { nameWithOwner }
        pageInfo {
          hasNextPage
          endCursor
        }
      }
    }
  }
'
```
from
https://cli.github.com/manual/gh_api

helayoty replied:
Added.

marosset commented:
Overall this looks good.
One question I had was is there a desire to save the query output to files instead of just saving the full json output to variables and then streaming that to the jq queries?

Priyankasaggu11929 commented:
Thanks so much for the review @marosset.

There is no specific reason to save the query output to files (and to be honest, I didn't try using variables).

However, saving the query output to files was useful when looping over the output to get multiple values, as in the [following case](https://github.com/kubernetes/sig-release/pull/1969/files#diff-bf9c8c5f61e3d0f8c973f2e1b9c24ba87a1191026f27471322f836da9f094c8fR138-R142): 

 ```
 for index in $(seq 0 $issues_loop_index_range);
do
	issue_id=$(jq ".data.repository.milestones.nodes[].issues.nodes[$index].id" "${MILESTONE_ISSUES_JSON}")
	milestone_issue_ids+=("${issue_id}")
done
 ```

marosset commented:
> Thanks so much for the review @marosset.
> 
> There is no specific reason to save the query output to files (and to be honest, I didn't try using variables).
> 
> However, saving the query output to files was useful when looping over the output to get multiple values, as in the [following case](https://github.com/kubernetes/sig-release/pull/1969/files#diff-bf9c8c5f61e3d0f8c973f2e1b9c24ba87a1191026f27471322f836da9f094c8fR138-R142):
> 
> ```
> for index in $(seq 0 $issues_loop_index_range);
> do
>    issue_id=$(jq ".data.repository.milestones.nodes[].issues.nodes[$index].id" "${MILESTONE_ISSUES_JSON}")
>    milestone_issue_ids+=("${issue_id}")
> done
> ```

If the raw json was saved as a variable things like
```bash
 issue_id=$(jq ".data.repository.milestones.nodes[].issues.nodes[$index].id" "${MILESTONE_ISSUES_JSON}")
```
can be updated to something like
```bash
 issue_id=$(jq ".data.repository.milestones.nodes[].issues.nodes[$index].id" <<< "$MILESTONE_ISSUES_JSON")
```
and then not much in the script would need to change.

Just a suggestion :)

helayoty commented:
@marosset  @mrbobbytables Appreciate your review so we can merge this PR and help the release team by automating the Bug Triage GH project :) 

Cc @hosseinsalahi @leonardpahlke

marosset commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,218 @@
+set -eu -o pipefail
+
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+REPO_ROOT="$(cd "${SCRIPT_DIR}/.." && pwd )"
```
It doesn't look like REPO_ROOT / TMPDIR are used now that the script is no longer writing files out to disk.
I think we can drop these variable assignments.

helayoty replied:
Done.

marosset commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,218 @@
+
+ORGANIZATION="kubernetes"
+REPOSITORY="kubernetes"
+PROJECT_NUMBER=${GITHUB_PROJECT_BETA_NUMBER}
```
Should we verify that PROJECT_NUMBER/MILESTONE are set (or assign a default value)?

marosset reviewed (commented):
A few small comments.
Overall looks good.

mrbobbytables reviewed (commented):
One FYI - on commit 1744554125a135b6410a7a48ce642f22ed51bb6f the leading @ needs to be removed from the commit message to remove the `invalid-commit-message` label.

I can take a closer look later, looping in some of the TL shadows for additional review as well

/assign @detiber @MadhavJivrajani

helayoty commented:
> One FYI - on commit [1744554](https://github.com/kubernetes/sig-release/commit/1744554125a135b6410a7a48ce642f22ed51bb6f) the leading @ needs to be removed from the commit message to remove the `invalid-commit-message` label.
> 
> I can take a closer look later, looping in some of the TL shadows for additional review as well
> 
> /assign @detiber @MadhavJivrajani

Fixed and squashed.

marosset commented:
/lgtm

detiber commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,228 @@
+            }' -f org=${ORGANIZATION} -f repo=${REPOSITORY} -f milestone="${MILESTONE}")"
+
+no_of_issues=$(jq ".data.repository.milestones.nodes[].issues.nodes | length" <<< "${MILESTONE_ISSUES_JSON}")
+issues_loop_index_range=$(( no_of_issues - 1 ))
```
In the case of pagination, `no_of_issues` does not return a single number, but rather a number for each paginated group.

For example, I ran a modified query to force pagination:
```sh
MILESTONE_ISSUE_JSON="$(gh api graphql --paginate -f query='
            query($org: String!, $repo: String!, $milestone: String!, $endCursor: String){
              repository(owner: $org, name: $repo) {
                description
                url
                milestones(states: [OPEN],first:1, query: $milestone) {
                  nodes{
                    issues(states:[OPEN, CLOSED], first: 5, after: $endCursor){
                      nodes{
                        id
                      }
                      pageInfo{
                        hasNextPage
                        endCursor
                      }
                    }
                  }
                }
              }
            }' -f org=kubernetes -f repo=kubernetes -f milestone="v1.24")"
```

This resulted in the following prettified output:
```json
{
  "data": {
    "repository": {
      "description": "Production-Grade Container Scheduling and Management",
      "url": "https://github.com/kubernetes/kubernetes",
      "milestones": {
        "nodes": [
          {
            "issues": {
              "nodes": [
                {
                  "id": "MDU6SXNzdWUzNTc5MzcyNzc="
                },
                {
                  "id": "MDU6SXNzdWU3ODU0NjQwMTQ="
                },
                {
                  "id": "MDU6SXNzdWU4Mjk1OTg1NDM="
                },
                {
                  "id": "MDU6SXNzdWU4MzMwNzU2Mzc="
                },
                {
                  "id": "MDU6SXNzdWU5Mzc0MzIxNjg="
                }
              ],
              "pageInfo": {
                "hasNextPage": true,
                "endCursor": "Y3Vyc29yOnYyOpLAzjfgFGg="
              }
            }
          }
        ]
      }
    }
  }
}
{
  "data": {
    "repository": {
      "description": "Production-Grade Container Scheduling and Management",
      "url": "https://github.com/kubernetes/kubernetes",
      "milestones": {
        "nodes": [
          {
            "issues": {
              "nodes": [
                {
                  "id": "MDU6SXNzdWU5MzgyNzE1Nzc="
                },
                {
                  "id": "MDU6SXNzdWU5NjYxMDUzNDM="
                },
                {
                  "id": "MDU6SXNzdWU5ODE2OTg2MzE="
                },
                {
                  "id": "I_kwDOAToIks48R-_H"
                },
                {
                  "id": "I_kwDOAToIks485qjC"
                }
              ],
              "pageInfo": {
                "hasNextPage": true,
                "endCursor": "Y3Vyc29yOnYyOpLAzjzmqMI="
              }
            }
          }
        ]
      }
    }
  }
}
{
  "data": {
    "repository": {
      "description": "Production-Grade Container Scheduling and Management",
      "url": "https://github.com/kubernetes/kubernetes",
      "milestones": {
        "nodes": [
          {
            "issues": {
              "nodes": [
                {
                  "id": "I_kwDOAToIks49mPsO"
                },
                {
                  "id": "I_kwDOAToIks4_Jl9m"
                },
                {
                  "id": "I_kwDOAToIks5DVQcj"
                },
                {
                  "id": "I_kwDOAToIks5EccUj"
                },
                {
                  "id": "I_kwDOAToIks5F5hqn"
                }
              ],
              "pageInfo": {
                "hasNextPage": true,
                "endCursor": "Y3Vyc29yOnYyOpLAzkXmGqc="
              }
            }
          }
        ]
      }
    }
  }
}
{
  "data": {
    "repository": {
      "description": "Production-Grade Container Scheduling and Management",
      "url": "https://github.com/kubernetes/kubernetes",
      "milestones": {
        "nodes": [
          {
            "issues": {
              "nodes": [
                {
                  "id": "I_kwDOAToIks5GL4xf"
                },
                {
                  "id": "I_kwDOAToIks5GMppR"
                },
                {
                  "id": "I_kwDOAToIks5GNW7c"
                },
                {
                  "id": "I_kwDOAToIks5GQSss"
                },
                {
                  "id": "I_kwDOAToIks5GRJuM"
                }
              ],
              "pageInfo": {
                "hasNextPage": true,
                "endCursor": "Y3Vyc29yOnYyOpLAzkZEm4w="
              }
            }
          }
        ]
      }
    }
  }
}
{
  "data": {
    "repository": {
      "description": "Production-Grade Container Scheduling and Management",
      "url": "https://github.com/kubernetes/kubernetes",
      "milestones": {
        "nodes": [
          {
            "issues": {
              "nodes": [
                {
                  "id": "I_kwDOAToIks5GRkQm"
                },
                {
                  "id": "I_kwDOAToIks5GZYK2"
                },
                {
                  "id": "I_kwDOAToIks5GaF2M"
                },
                {
                  "id": "I_kwDOAToIks5GkPkh"
                },
                {
                  "id": "I_kwDOAToIks5GmTJS"
                }
              ],
              "pageInfo": {
                "hasNextPage": true,
                "endCursor": "Y3Vyc29yOnYyOpLAzkaZMlI="
              }
            }
          }
        ]
      }
    }
  }
}
{
  "data": {
    "repository": {
      "description": "Production-Grade Container Scheduling and Management",
      "url": "https://github.com/kubernetes/kubernetes",
      "milestones": {
        "nodes": [
          {
            "issues": {
              "nodes": [
                {
                  "id": "I_kwDOAToIks5Gtduc"
                },
                {
                  "id": "I_kwDOAToIks5Gvivj"
                },
                {
                  "id": "I_kwDOAToIks5GxtNC"
                },
                {
                  "id": "I_kwDOAToIks5G48Xk"
                },
                {
                  "id": "I_kwDOAToIks5G-4-C"
                }
              ],
              "pageInfo": {
                "hasNextPage": true,
                "endCursor": "Y3Vyc29yOnYyOpLAzkb7j4I="
              }
            }
          }
        ]
      }
    }
  }
}
{
  "data": {
    "repository": {
      "description": "Production-Grade Container Scheduling and Management",
      "url": "https://github.com/kubernetes/kubernetes",
      "milestones": {
        "nodes": [
          {
            "issues": {
              "nodes": [
                {
                  "id": "I_kwDOAToIks5HBDuK"
                },
                {
                  "id": "I_kwDOAToIks5IhDUu"
                }
              ],
              "pageInfo": {
                "hasNextPage": false,
                "endCursor": "Y3Vyc29yOnYyOpLAzkiENS4="
              }
            }
          }
        ]
      }
    }
  }
}
```

Which when filtered through jq for length results in:

```sh
> echo "${MILESTONE_ISSUE_JSON}" | jq ".data.repository.milestones.nodes[].issues.nodes | length"
5
5
5
5
5
5
2
```

I'm not sure if there is a way within jq to produce the proper length, but you could always wrap this with an outer loop that consumes each line of the length output separately

Priyankasaggu11929 replied:
Thank you so much for including the example above, @detiber 

(it actually forced me to try the "forced pagination" myself, and only then was I able to understand the issue and proceed.) 🙂 

I've addressed the changes introduced by the pagination in the script.

detiber commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,228 @@
+            }' -f org=${ORGANIZATION} -f repo=${REPOSITORY} -f milestone="${MILESTONE}")"
+
+no_of_prs=$(jq ".data.repository.milestones.nodes[].pullRequests.nodes | length" <<< "${MILESTONE_PRS_JSON}")
+pr_loop_index_range=$(( no_of_prs - 1 ))
```
Similar to above, the length calculation here would break for paginated output.

Priyankasaggu11929 replied:
addressed. Thank you!

detiber commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,228 @@
+# Add data into the Project Board
+echo  '[INFO] Feeding data from k/k issues into the GitHub Project Beta'
+
+for index in $(seq 0 ${issues_loop_index_range});
```
This would also need to account for the pagination issue I mentioned above.

Priyankasaggu11929 replied:
addressed. Thank you!

detiber commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,228 @@
+
+echo  '[INFO] Feeding data from k/k PRs into the GitHub Project Beta' 
+
+for index in $(seq 0 ${pr_loop_index_range});
```
This would also need to account for the pagination issue I mentioned above.

detiber commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,228 @@
+
+for index in $(seq 0 $issues_loop_index_range);
+do
+	issue_id=$(jq ".data.repository.milestones.nodes[].issues.nodes[$index].id" <<< "${MILESTONE_ISSUES_JSON}")
```
I believe this would need to take into account indexing for the root object when receiving paginated responses as well. I'm not quite sure how to do that in jq off the top of my head, though.

Priyankasaggu11929 replied:
addressed. Thank you! Did some hacky way, but takes care of the pagination well.

detiber commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,228 @@
+
+for index in $(seq 0 $pr_loop_index_range);
+do
+  pr_id=$(jq ".data.repository.milestones.nodes[].pullRequests.nodes[$index].id" <<< "${MILESTONE_PRS_JSON}")
```
I believe this would need to take into account indexing for the root object when receiving paginated responses as well. I'm not quite sure how to do that in jq off the top of my head, though.

Priyankasaggu11929 replied:
addressed. Thank you!

detiber reviewed (commented):
Overall, this is looking great to me, however I believe there are still a few changes needed to properly support paginated results returned by the GitHub GraphQL API.

Priyankasaggu11929 commented:
Thanks so much for the review, @detiber.

I will spend a bit more time on the suggested pagination points and accordingly the jq bits as well. Will get back. 🙂

mrbobbytables commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,228 @@
+
+# function definitions
+function get_field_ids_from_github_beta_project() {
+    if [ $1 == "project-id" ]
```
```suggestion
    if [ "$1" == "project-id" ]
```
nit: while technically probably not needed, in general it's good practice to double quote variables. This can be applied to the other checks below

Priyankasaggu11929 replied:
addressed. Thank you!

mrbobbytables commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,228 @@
+                        }
+                        }
+                    }
+                    }' -f org=${ORGANIZATION} -F number=${PROJECT_NUMBER} --jq "$query")"
```
```suggestion
                    }' -f org="${ORGANIZATION}" -F number="${PROJECT_NUMBER}" --jq "$query")"
```

Adding double quotes to the other variables 👍

Priyankasaggu11929 replied:
addressed. Thank you!

mrbobbytables commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,228 @@
+                  }
+                }
+              }
+            }' -f org=${ORGANIZATION} -f repo=${REPOSITORY} -f milestone="${MILESTONE}")"
```
```suggestion
            }' -f org="${ORGANIZATION}" -f repo="${REPOSITORY}" -f milestone="${MILESTONE}")"
```
Adding double quotes to the other vars.

Priyankasaggu11929 replied:
addressed. Thank you!

mrbobbytables commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,228 @@
+                        }
+                    }
+                    }' -f org=${ORGANIZATION} -F number=${PROJECT_NUMBER} --jq "$query")"
+    echo $ID
```
```suggestion
    echo "$ID"
```

Priyankasaggu11929 replied:
added quotes to this and all following similar blocks. Thank you for pointing.

Priyankasaggu11929 commented:
@helayoty I've addressed all the review comments above.

The updated script is here ~ https://github.com/Priyankasaggu11929/sig-release/blob/sync-bug-triage-github-project/release-team/hack/sync-bug-triage-github-project-beta.sh

Need your help to refresh the PR commit with the changes ^ for another round of review. Thank you! 🙂

Priyankasaggu11929 commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,228 @@
+                  }
+                }
+              }
+            }' -f org=${ORGANIZATION} -f repo=${REPOSITORY} -f milestone="${MILESTONE}")"
```
addressed. Thank you!

Priyankasaggu11929 commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,228 @@
+                  }
+                }
+              }
+            }' -f org=${ORGANIZATION} -f repo=${REPOSITORY} -f milestone="${MILESTONE}")"
```
addressed. Thank you!

Priyankasaggu11929 commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,228 @@
+
+set -eu -o pipefail
+
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
```
addressed. Thank you!

Priyankasaggu11929 commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,228 @@
+              }
+            }' -f project="${PROJECT_ID}" -f issue="${milestone_issue_ids[index]}" --jq '.data.addProjectV2ItemById.item.id')"
+
+    $(add_items_to_github_beta_project "${item_id}" "${ISSUE_OPTION_ID}")           
```
Removed the braces. Thank you for pointing.

Priyankasaggu11929 commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,228 @@
+                  }
+                }
+              }
+            }' -f org=${ORGANIZATION} -f repo=${REPOSITORY} -f milestone="${MILESTONE}")"
```
Added quotes. Thank you!

detiber commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,232 @@
+# limitations under the License.
+
+# Usage: `sync-bug-triage-github-project-beta.sh`
+# Pre-requisite: add a GITHUB_TOKEN as the environment variable (Ref: https://github.com/kubernetes/org/issues/3558)
```
It would probably be good to add a check to verify that `$GITHUB_TOKEN` is set at runtime, similar to how `$PROJECT_NUMBER` and `$MILESTONE` are verified below.

Priyankasaggu11929 replied:
added a verify check for the GITHUB_TOKEN variable ([here](https://github.com/Priyankasaggu11929/sig-release/blob/sync-bug-triage-github-project/release-team/hack/sync-bug-triage-github-project-beta.sh))

Will update the PR commits soon, with the same.

detiber commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,232 @@
+
+set -eu -o pipefail
+
+ORGANIZATION="kubernetes"
+REPOSITORY="kubernetes"
```
These could be made overridable, while defaulting to "kubernetes":

```suggestion
: "${ORGANIZATION:=kubernetes}"
: "${REPOSITORY:=kubernetes}"
```

`:` means do nothing, and the `"${ORGANIZATION:=kubernetes}"` uses parameter expansion to set ORGANIZATION to "kubernetes" when it is undefined or an empty string.

Priyankasaggu11929 replied:
Thanks so much @detiber. Have address the new review comments [here](https://github.com/Priyankasaggu11929/sig-release/blob/sync-bug-triage-github-project/release-team/hack/sync-bug-triage-github-project-beta.sh).

Will refresh the PR commits with the same soon.

detiber commented on release-team/hack/sync-bug-triage-github-project-beta.sh:
```diff
@@ -0,0 +1,232 @@
+ORGANIZATION="kubernetes"
+REPOSITORY="kubernetes"
+
+PROJECT_NUMBER=${GITHUB_PROJECT_BETA_NUMBER:-}
+MILESTONE=${RELEASE_MILESTONE:-}
+
+if [ "$PROJECT_NUMBER" == "" ]; then
+      echo "[Error] Required environment variable \"GITHUB_PROJECT_BETA_NUMBER\" is not set."
+      exit 1
+fi
+
+if [ "$MILESTONE" == "" ]; then
+      echo "[Error] Required environment variable \"RELEASE_MILESTONE\" is not set."
+      exit 1
+fi
```
If you wanted to, you could eliminate the intermediate variables here and have the user supply `$PROJECT_NUMBER` and `$MILESTONE` directly:
```suggestion
if [ "${PROJECT_NUMBER:-}" == "" ]; then
      echo "[Error] Required environment variable \"PROJECT_NUMBER\" is not set."
      exit 1
fi

if [ "${MILESTONE:-}" == "" ]; then
      echo "[Error] Required environment variable \"MILESTONE\" is not set."
      exit 1
fi
```

Priyankasaggu11929 replied:
addressed. Thank you for the review. :)

detiber reviewed (commented):
Minor optional suggestions.

/lgtm

detiber commented:
/lgtm

mrbobbytables reviewed (commented):
/lgtm

leonardpahlke commented:
Thank you @helayoty, @Priyankasaggu11929 and everyone else!! 

/approve
